prql-compiler/src/error.rs

```diff
@@ -97,115 +111,92 @@ impl Display for Error {
-        source_id: &str,
-        source: Source,
-        color: bool,
-    ) -> (String, String);
-}
+pub fn downcast(error: anyhow::Error) -> ErrorMessage {
+    let mut span = None;
+    let mut hint = None;
 
-impl IntoErrorMessage for anyhow::Error {
-    /// Convert error into human-readable message and error location.
-    fn into_error_message(self, source_id: &str, source: &str, color: bool) -> ErrorMessage {
-        let source = Source::from(source);
-        let location = self.to_location(&source);
+    let error = match error.downcast::<ErrorMessage>() {
+        Ok(message) => return message,
+        Err(error) => error,
+    };
 
-        let (line, output) = self.into_error_message_and_output(source_id, source, color);
+    let reason = match error.downcast::<Error>() {
+        Ok(error) => {
+            span = error.span;
+            hint = error.help;
 
-        ErrorMessage {
-            message: output,
-            line,
-            location,
+            error.reason.message()
         }
-    }
-    fn to_location(&self, source: &Source) -> Option<SourceLocation> {
-        let span = if let Some(error) = self.downcast_ref::<Error>() {
-            if let Some(span) = error.span {
-                Range::from(span)
-            } else {
-                return None;
+        Err(error) => {
+            match error.downcast::<PestError>() {
+                Ok(error) => {
+                    let range = pest::as_range(&error);
+                    span = Some(Span {
+                        start: range.start,
+                        end: range.end,
+                    });
+
+                    pest::as_message(&error)
+                }
+                Err(error) => {
+                    // default to basic Display
+                    format!("{:#?}", error)
+                }
             }
-        } else if let Some(error) = self.downcast_ref::<PestError>() {
-            pest::as_range(error)
-        } else {
-            return None;
-        };
-
-        let start = source.get_offset_line(span.start)?;
-        let end = source.get_offset_line(span.end)?;
-
-        Some(SourceLocation {
-            start: (start.1, start.2),
-            end: (end.1, end.2),
-        })
+        }
+    };
+
+    ErrorMessage {
+        reason,
+        hint,
+        span,
+        display: None,
+        location: None,
     }
-    // TODO: should we consolidate between `into_error_message` and `into_error_message_and_output`?
-    fn into_error_message_and_output(
-        self,
-        source_id: &str,
-        source: Source,
-        color: bool,
-    ) -> (String, String) {
-        let config = Config::default().with_color(color);
-
-        if let Some(error) = self.downcast_ref::<Error>() {
-            let message = error.reason.message();
+}
 
-            if let Some(span) = error.span {
-                let span = Range::from(span);
+impl ErrorMessage {
+    pub fn as_json(&self) -> String {
```

Comment:
FWIW technically I think this should be `to_json` based on https://rust-lang.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv, but no huge deal

Reply:
I didn't know there is actually a guide on that! I'll change it.